--- zlib-1.2.11/contrib/s390/dfltcc_deflate.h.old	2023-05-04 09:39:23.423753908 +0000
+++ zlib-1.2.11/contrib/s390/dfltcc_deflate.h	2023-05-04 09:39:48.087753908 +0000
@@ -45,11 +45,11 @@ int ZLIB_INTERNAL dfltcc_deflate_get_dic
 #define DEFLATE_DONE dfltcc_deflate_done
 #define DEFLATE_BOUND_ADJUST_COMPLEN(strm, complen, source_len) \
     do { \
-        if (dfltcc_can_deflate((strm))) \
+	if (deflateStateCheck((strm)) || dfltcc_can_deflate((strm))) \
             (complen) = DEFLATE_BOUND_COMPLEN(source_len); \
     } while (0)
 #define DEFLATE_NEED_CONSERVATIVE_BOUND(strm) (dfltcc_can_deflate((strm)))
 #define DEFLATE_HOOK dfltcc_deflate
 #define DEFLATE_NEED_CHECKSUM(strm) (!dfltcc_can_deflate((strm)))
 
-#endif
\ No newline at end of file
+#endif
